orcid-core/src/main/java/org/orcid/core/utils/activities/ActivitiesGroupGenerator.java

```diff
@@ -17,50 +17,66 @@
 package org.orcid.core.utils.activities;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
+import org.orcid.jaxb.model.record_rc2.GroupAble;
 import org.orcid.jaxb.model.record_rc2.GroupableActivity;
 
 public class ActivitiesGroupGenerator {    
 
     private List<ActivitiesGroup> groups = new ArrayList<ActivitiesGroup>();
     
+    private Map<GroupAble, ActivitiesGroup> lookup = new HashMap<GroupAble, ActivitiesGroup>();
+    
     
     public void group(GroupableActivity activity) {
         if(groups.isEmpty()) {
             //If it is the first activity, create a new group for it
             ActivitiesGroup newGroup = new ActivitiesGroup(activity);
             groups.add(newGroup);
+            for (GroupAble g :newGroup.getGroupKeys()){
+                lookup.put(g, newGroup);
+            }
         } else {            
             //If it is not the first activity, check which groups it belongs to
-            List<Integer> belongsTo = new ArrayList<Integer>();
-            for(int i = 0; i < groups.size(); i++) {
-                ActivitiesGroup group = groups.get(i);
-                if(group.belongsToGroup(activity)) {
-                    belongsTo.add(i);
-                }
+            List<ActivitiesGroup> belongsTo = new ArrayList<ActivitiesGroup>();
+            ActivitiesGroup thisGroup = new ActivitiesGroup(activity);
+            for (GroupAble g :thisGroup.getGroupKeys()){
```

Comment:
So we have a group with 150,000 works with the same key + a unique key what happens?